### PR TITLE
UML-825 fix input acceptance of viewer and actor codes

### PR DESCRIPTION
**Pull request by williamfalconeruk**
## Purpose
To fix issues with actor does where no prefix is given but the first letter is a V for viewer and C for actor, and it strips off the character, resulting in errors in finding the code.

Fixes UML-825.

## Approach
- expand known test case acceptable codes that can be entered
- find a good combination of regex to accept the codes
- cover these with additional test scenarios

**Note:** there are 2 Todos that have been covered by UML-831 for someone to tackle, where we create a custom filter to sanitize the input further. see https://github.com/laminas/laminas-filter/blob/master/src/AbstractFilter.php 

## Learning
- regex named groups and backreferences using \g: https://www.regular-expressions.info/named.html
- filters in PHP laminas forms

## Checklist

* [X] I have performed a self-review of my own code
* [ ] I have added relevant logging with appropriate levels to my code
* [ ] I have updated documentation (Confluence/GitHub wiki/tech debt doc) where relevant
* [X] I have added tests to prove my work
* [ ] The product team have tested these changes

### CODE REVIEW

File: service-front/app/src/Viewer/src/Handler/EnterCodeHandler.php [hunk 1]
```diff
@@ -37,12 +37,13 @@ public function handle(ServerRequestInterface $request): ResponseInterface
 
             if ($form->isValid()) {
                 $lpaCode = $form->getData()['lpa_code'];
-
+                //TODO: refactor this to a custom filter on the form. Ticket UML-831
+                $lpaCode = str_replace(' ', '', $lpaCode);
                 // Remove V from start of the code if present
-                $lpaCode = preg_replace('/^[v]?/i', '', $lpaCode);
+                $lpaCode = preg_replace('/^(V(?\'dash\'-| - ))?/i', '', $lpaCode);
```
Comment (cooperaj) [inline service-front/app/src/Viewer/src/Handler/EnterCodeHandler.php:43]: To save escaping the quotes " ' " in the regex wrap it in doublequotes ' " '

Comment (williamfalconeruk) [reply service-front/app/src/Viewer/src/Handler/EnterCodeHandler.php:43]: cool - will do

File: service-front/app/features/context/UI/AccountContext.php [hunk 2]
```diff
@@ -556,6 +556,14 @@ public function iAmOnTheAddAnLPAPage()
      */
     public function iRequestToAddAnLPAWithValidDetailsUsing(string $code)
     {
+
+        // this is to give what is the expected match.
```
[the remark this answers is not in the document]
Comment (williamfalconeruk) [reply service-front/app/features/context/UI/AccountContext.php:560]: true - could change that.

File: service-front/app/features/context/UI/ViewerContext.php [hunk 3]
```diff
@@ -82,16 +85,31 @@ public function iHaveBeenGivenAccessToAnExpiredLPAViaShareCode() {
...
         $this->ui->assertElementContainsText('a[role=button]', 'Start');
         $this->ui->clickLink('Start');
     }
 
+    /**
+     * @When /^I give a valid LPA share code of (.*)$/
+     */
+    public function iGiveAValidLPAShareCodeOf($code)
+    {
+        $lpaCode = str_replace(' ', '', $code);
```
[the remark this answers is not in the document]
Comment (williamfalconeruk) [reply service-front/app/features/context/UI/ViewerContext.php:100]: as above

File: service-front/app/features/view-lpa-using-sharecode.feature [hunk 4]
```diff
@@ -37,6 +37,39 @@ Feature: View an LPA via sharecode
...
 
   @ui
+  Scenario Outline: The user is able to enter a valid LPA code
+    Given I have been given access to an LPA via share code
+    And I access the viewer service
+    When I give a valid LPA share code of <accessCode>
+    And I confirm the LPA is correct
+    Then I can see the full details of the valid LPA
+    Examples:
+    |accessCode             |
```
[the remark this answers is not in the document]
Comment (williamfalconeruk) [reply service-front/app/features/view-lpa-using-sharecode.feature:47]: I wanted to cover as much as possible, to ensure i'd hadn't missed anything. they run really quickly anyway....

File: service-front/app/src/Actor/src/Form/LpaAdd.php [hunk 5]
```diff
@@ -76,7 +76,7 @@ public function getInputFilterSpecification(): array
                     [
                         'name'    => Regex::class,
                         'options' => [
-                            'pattern' => "/(^[cC]?|^)(?'dash' |-|){0,6}[[:alnum:]]{4}(\g'dash'){0,6}[[:alnum:]]{4}(\g'dash'){0,6}[[:alnum:]]{4}$/",
+                            'pattern' => "/^(C(?'dash'-| - ))?[[:alnum:]]{4}(\g'dash')?[[:alnum:]]{4}(\g'dash')?[[:alnum:]]{4}$/i",
```
[the remark this answers is not in the document]
Comment (williamfalconeruk) [reply service-front/app/src/Actor/src/Form/LpaAdd.php:79]: i've checked the formats with katie on this, dashes with spaces are allowed as are dashes. will confirm the spaces though...

[the remark this answers is not in the document]
Comment (williamfalconeruk) [reply service-front/app/src/Actor/src/Form/LpaAdd.php:79]: so far I have: 
- ` <spc>-<spc>` 
-  `-` no <spc>
 - no gaps

File: service-front/app/features/actor-add-an-lpa.feature [hunk 6]
```diff
@@ -24,11 +24,23 @@ Feature: Add an LPA
     And The LPA is successfully added
 
     Examples:
-      | passcode       |
-      | xyuphwqrechv   |
-      | XYUPHWQRECHV   |
-      | XYUP-hwqr-EcHv |
-      | C - XYUP - hwqr - Echv |
+      | passcode               |
```
[the remark this answers is not in the document]
Comment (williamfalconeruk) [reply service-front/app/features/actor-add-an-lpa.feature:27]: being over thorough i guess (wanted to be sure)... speed of this is quick though...

[the remark this answers is not in the document]
Comment (williamfalconeruk) [reply service-front/app/features/actor-add-an-lpa.feature:27]: thinking about it - end users might not mix their case if they copy and paste, I suppose when they enter the code they might do this though....

File: service-front/app/src/Viewer/src/Handler/EnterCodeHandler.php (hunk 1 above)
Comment (williamfalconeruk) [reply service-front/app/src/Viewer/src/Handler/EnterCodeHandler.php:43]: actually need to fix this slightly - as it doesn't need all of that in it, at this point.

File: service-front/app/src/Actor/src/Form/LpaAdd.php (hunk 5 above)
[the remark this answers is not in the document]
Comment (williamfalconeruk) [reply service-front/app/src/Actor/src/Form/LpaAdd.php:79]: ok so had some clarification from @Raffers - will make the changes to include spaces...

File: service-front/app/features/actor-add-an-lpa.feature (hunk 6 above)
[the remark this answers is not in the document]
Comment (cooperaj) [reply service-front/app/features/actor-add-an-lpa.feature:27]: As long as the tests are quick it's ok I would think.

[the remark this answers is not in the document]
Comment (williamfalconeruk) [reply service-front/app/features/actor-add-an-lpa.feature]: ok tried to rationalise these - but ended up added more tests, the net effect is fewer scenarios though...

File: service-front/app/src/Actor/src/Form/LpaAdd.php [hunk 7]
```diff
@@ -76,7 +76,7 @@ public function getInputFilterSpecification(): array
                     [
                         'name'    => Regex::class,
                         'options' => [
-                            'pattern' => "/(^[cC]?|^)(?'dash' |-|){0,6}[[:alnum:]]{4}(\g'dash'){0,6}[[:alnum:]]{4}(\g'dash'){0,6}[[:alnum:]]{4}$/",
+                            'pattern' => "/^(C(?'dash'-| ){1,6})?[[:alnum:]]{4}(\g'dash'){0,6}[[:alnum:]]{4}(\g'dash'){0,6}[[:alnum:]]{4}$/i",
```
Comment (GemTay) [inline service-front/app/src/Actor/src/Form/LpaAdd.php:79]: Looks like theres an accidental "i" on the end of the regex expression

File: service-front/app/features/context/UI/AccountContext.php [hunk 8]
```diff
@@ -550,10 +550,13 @@ public function iAmOnTheAddAnLPAPage()
...
     /**
-     * @When /^I request to add an LPA with valid details using (.*)$/
+     * @When /^I request to add an LPA with valid details using (.*) which matches (.*)$/
      */
-    public function iRequestToAddAnLPAWithValidDetailsUsing(string $code)
+    public function iRequestToAddAnLPAWithValidDetailsUsing(string $code, string $storedCode)
     {
+
+
+
```
Comment (GemTay) [inline service-front/app/features/context/UI/AccountContext.php:559]: Couple of blank lines to remove

**Pull request merged**